lucetc/src/module.rs

```diff
@@ -401,4 +404,14 @@ impl<'a> ModuleEnvironment<'a> for ModuleInfo<'a> {
         }
         Ok(())
     }
+
+    fn declare_func_name(&mut self, func_index: FuncIndex, name: &'a str) -> WasmResult<()> {
+        let unique_func_index = *self
+            .function_mapping
+            .get(func_index)
+            .expect("function indices are valid");
+        self.function_names
+            .insert(unique_func_index, name.to_owned());
+        Ok(())
+    }
```

Comment:
Over in bytecodealliance/cranelift#811, we added a method to the `ModuleEnvironment` trait for handling function names found in the name section. We define that here. The main point of complexity here is that we need to get the unique function id using `self.function_mapping`, or we won't have an easy way to get the name later.

Reply:
Note: Above I've defined `ModuleInfo::function_names` as `HashMap<UniqueFuncIndex, String>`. I _wanted_ a way to use a `PrimaryMap` but I'm not sure that's possible here. Let me know if I've missed something w/ respect to that.

Reply:
I think the thing you'd end up needing is in fact a `SecondaryMap`, but with a `SecondaryMap<UniqueFunctionIndex, &'a str>` I'm not sure how the lifetime of its `default` would work (maybe `SecondaryMap::with_default("")` does actually work? that would be super cool). We could probably do the same `HashMap` -> `SecondaryMap` treatment for `function_bodies` as well, but I'm not sure if it's very meaningful.

if memory serves, `PrimaryMap` is intended be kind of an authoritative source for the keyspace, so two `PrimaryMap` keyed on the same thing seem like they'd be in error. I see we have a few of those in `module.rs` and dunno if that was an oversight, intentional (and my understanding wrong), or `SecondaryMap` didn't exist yet?

Reply:
https://github.com/cranestation/cranelift/blob/3217e4e3253bd15a93f1a191345c5e3a0dab31de/upstream#L22-L23

> There should only be a single `PrimaryMap` instance for a given `EntityRef` type, otherwise
> conflicting references will be created. Using unknown keys for indexing will cause a panic.

It seems like you are right, re: `PrimaryMap` v. `SecondaryMap`. And it turns out the default lifetime worked out just fine! :tada: